server/app/scanpy_engine/scanpy_engine.py

```diff
@@ -166,11 +167,58 @@ def _validate_and_initialize(self):
         self._alias_annotation_names(Axis.OBS, self.config["obs_names"])
         self._alias_annotation_names(Axis.VAR, self.config["var_names"])
         self._validate_data_types()
-        self._validate_data_calculations()
         self.cell_count = self.data.shape[0]
         self.gene_count = self.data.shape[1]
+        self._default_and_validate_layouts()
         self._create_schema()
 
+    @requires_data
+    def _default_and_validate_layouts(self):
+        """ function:
+            a) generate list of default layouts, if not already user specified
+            b) validate layouts are legal.  remove/warn on any that are not
+            c) cap total list of layouts at global const MAX_LAYOUTS
+        """
+        layouts = self.config['layout']
+        # handle default
+        if layouts is None or len(layouts) == 0:
+            # load default layouts from the data.
+            layouts = [key[2:] for key in self.data.obsm_keys() if type(key) == str and key.startswith("X_")]
+            if len(layouts) == 0:
+                raise PrepareError(
+                    f"Unable to find precomputed layouts within the dataset."
+                    f"You can run `cellxgene prepare --layout {self.config['layout']} <datafile>` "
```

Comment:
I don't want to push `prepare` too hard, could we add in the text about calculating a layout and saving it back to the h5ad too?

Reply:
This text is copied from a similar warning/error in the layout route :-)

I think that adequate explanation may require more prose than comfortably fits in a warning.   I think we should add this to a data prep document in end user docs (a long outstanding need).  

I will simply remove the text.